src/dtls.c

```diff
@@ -62,14 +71,278 @@ void DtlsResetState(WOLFSSL *ssl)
+        return BUFFER_ERROR;
+
+    ch->pv = (ProtocolVersion*)(input + idx);
+    idx += OPAQUE16_LEN;
+    ch->random = (byte*)(input + idx);
+    idx += RAN_LEN;
+    idx += ReadVector8(input + idx, &ch->sessionId);
+    if (idx > helloSz)
+        return BUFFER_ERROR;
+    if (helloSz - idx < OPAQUE8_LEN)
+        return BUFFER_ERROR;
+    idx += ReadVector8(input + idx, &ch->cookie);
+    if (idx > helloSz)
+        return BUFFER_ERROR;
+    if (helloSz - idx < OPAQUE16_LEN)
+        return BUFFER_ERROR;
+    idx += ReadVector16(input + idx, &ch->cipherSuite);
+    if (idx > helloSz)
+        return BUFFER_ERROR;
+    idx += ReadVector8(input + idx, &ch->compression);
+    if (idx > helloSz)
+        return BUFFER_ERROR;
+    if (helloSz - idx < OPAQUE16_LEN)
+        return BUFFER_ERROR;
+    idx += ReadVector16(input + idx, &ch->extension);
+    if (idx > helloSz)
+        return BUFFER_ERROR;
+    ch->length = idx;
+    return 0;
+}
+
+#ifdef WOLFSSL_DTLS_NO_HVR_ON_RESUME
+static int TlsxFindByType(WolfSSL_Vector* ret, word16 extType,
+    WolfSSL_Vector exts)
+{
+    word32 len, idx = 0;
+    word16 type;
+    WolfSSL_Vector ext;
+
+    XMEMSET(ret, 0, sizeof(*ret));
+    len = exts.size;
+    /* type + len */
+    while (len >= OPAQUE16_LEN + OPAQUE16_LEN) {
+        ato16(exts.elements + idx, &type);
+        idx += OPAQUE16_LEN;
+        idx += ReadVector16(exts.elements + idx, &ext);
+        if (idx > exts.size)
+            return BUFFER_ERROR;
+        if (type == extType) {
+            XMEMCPY(ret, &ext, sizeof(ext));
+            return 0;
+        }
+        len = exts.size - idx;
+    }
+    return 0;
+}
+
+#ifdef HAVE_SESSION_TICKET
+static int TlsTicketIsValid(WOLFSSL* ssl, WolfSSL_Vector exts, byte* isValid)
+{
+    byte tempTicket[SESSION_TICKET_LEN];
+    InternalTicket* it;
+    WolfSSL_Vector tlsxSessionTicket;
+    int ret;
+
+    *isValid = 0;
+    ret = TlsxFindByType(&tlsxSessionTicket, TLSX_SESSION_TICKET, exts);
+    if (ret != 0)
+        return ret;
+    if (tlsxSessionTicket.size == 0)
+        return 0;
+    if (tlsxSessionTicket.size > SESSION_TICKET_LEN)
+        return 0;
+    XMEMCPY(tempTicket, tlsxSessionTicket.elements, tlsxSessionTicket.size);
+    ret = DoDecryptTicket(ssl, tempTicket, (word32)tlsxSessionTicket.size, &it);
+    if (ret != WOLFSSL_TICKET_RET_OK && ret != WOLFSSL_TICKET_RET_CREATE)
+        return 0;
+    ForceZero(it, sizeof(InternalTicket));
+    *isValid = 1;
+    return 0;
+}
+#endif /* HAVE_SESSION_TICKET */
+
+static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_Vector sessionID,
+    byte* isValid)
+{
+    *isValid = 0;
+    if (ssl->options.sessionCacheOff)
+        return 0;
+    if (sessionID.size != ID_LEN)
+        return 0;
+#ifdef HAVE_EXT_CACHE
+    {
+        WOLFSSL_SESSION* sess;
+
+        if (ssl->ctx->get_sess_cb != NULL) {
+            int unused;
+            sess =
+                ssl->ctx->get_sess_cb(ssl, sessionID.elements, ID_LEN, &unused);
```

Comment:
wolfSSL_GetSessionFromCache checks whether both are for TLS 1.3 or both not.
May be extract the code from wolfSSL_GetSessionFromCache and have this function call it.

Reply:
I think this can over-complicate things without any valuable gain.
If the version mismatch the CH will be rejected by the stateful code later.
Doing that check here implies we should do also version negotiation statelessly, but this is something I would like to avoid. 
What do you think?

Reply:
OK. No protocol negotiation done yet, so can't compare.
Leave as it is.